File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {
    public int solution(int[][] jobs) {
        PriorityQueue<Integer[]> pq = new PriorityQueue<>((job1, job2) -> job1[1] - job2[1]);
        int totalWaitingTime = 0;
        int realtime = 0;
        int idx = 0;
        
        Arrays.sort(jobs, (job1, job2) -> {
            return job1[0] == job2[0]
                ? job1[1] - job2[1]
                : job1[0] - job2[0];
        });
        
        while (true) {
            while (idx < jobs.length
                   && jobs[idx][0] < realtime) {
                Integer[] job = {jobs[idx][0], jobs[idx][1]};
                pq.add(job);
                idx++;
            }
            if (pq.isEmpty()) {
                if (idx >= jobs.length) {
                    break;
                }
                int[] job = jobs[idx++];
                realtime = job[0] + job[1];
                totalWaitingTime += job[1];
                continue;
            }
            Integer[] job = pq.poll();
            totalWaitingTime += realtime + job[1] - job[0];
            System.out.println(job[1]); 
            System.out.println(realtime + job[1] - job[0]);
            realtime += job[1];
        }
        
        return totalWaitingTime / jobs.length;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.kt
import java.util.PriorityQueue

data class Job(val jobIndex: Int, val requestTime: Int, val workingTime: Int)

class Solution {
    private val queue = PriorityQueue<Job>(compareBy({it.workingTime}, {it.requestTime}, {it.jobIndex}))
    private var currentTime = 0
    private var totalReturnTime = 0
    
    fun solution(jobData: Array<IntArray>): Int {
        val jobs = jobData
            .withIndex()
            .map {Job(it.index, it.value[0], it.value[1])}
            .sortedBy {it.requestTime}
        
        var index = 0
        
        while (index < jobs.size) {
            if (queue.isEmpty() && jobs[index].requestTime > currentTime) {
                currentTime = jobs[index].requestTime
            }
            
            while (index < jobs.size 
                   && jobs[index].requestTime <= currentTime) {
                queue.add(jobs[index++])
            }
            
            process()
        }
        
        while (queue.isNotEmpty()) {
            process()
        }
        
        return (totalReturnTime / jobs.size)
    }
    
    fun process() {
        if (queue.isEmpty()) {
            return
        }
        val job = queue.remove()
        currentTime += job.workingTime
        totalReturnTime += (currentTime - job.requestTime)
    }
}